split-webapp/split/src/pages/Split/index.jsx

```diff
@@ -0,0 +1,388 @@
+            [userId]: {
+              ...transaction.users.byId[userId],
+              name: newName
+            }
+          }
+        }
+      }
+    });
+  };
+
+  handlePayeeChange = userId => {
+    const { transaction } = this.state;
+
+    this.setState({
+      transaction: { ...transaction, payed: userId }
+    });
+  };
+
+  handleTitleChange = event => {
+    const { transaction } = this.state;
+    const title = event.target.value;
+
+    this.setState({
+      transaction: { ...transaction, title }
+    });
+  };
+
+  handleTotalChange = (event, cost) => {
+    const { transaction } = this.state;
+
+    this.setState({
+      transaction: { ...transaction, cost }
+    });
+  };
+
+  normalizeInput = () => {
+    this.setState({
+      transaction: this.getNormalizedTransaction()
+    });
+  };
+
+  split = async () => {
+    const transaction = this.getNormalizedTransaction();
+    const { history } = this.props;
+    const { users, cost } = transaction;
+    const perPersonCost = cost / users.allIds.length;
+    const title = transaction.title || DEFAULT_BILL_TITLE;
+
+    const paymentArray = [];
+
+    // Show visual feedback that the user's intention of submitting
+    // the form has been acknowledged.
+    this.setState({
+      showErrors: false,
+      submitAcknowledged: true
+    });
+    await new Promise(resolve => setTimeout(resolve, 200));
+
+    if (this.validate().hasError) {
+      // Begin live-validations after first failed submission.
+      this.setState({
+        showErrors: true,
+        submitAcknowledged: false
+      });
+      return;
+    }
+
+    for (const userId of users.allIds) {
+      if (userId !== transaction.payed) {
+        paymentArray.push({
+          person: users.byId[userId].name,
+          amount: perPersonCost
+        });
+      }
+    }
+
+    const bill = {
+      title,
+      payer: transaction.users.byId[transaction.payed].name,
+      total: cost,
+      outstanding_payments: paymentArray
+    };
+
+    createBill(bill);
+
+    history.push("/home/transactions");
+  };
+
+  validate() {
+    const transaction = this.getNormalizedTransaction();
+
+    const errors = {};
+
+    if (transaction.cost === 0) {
+      errors.cost = "Cannot split a zero amount";
+      errors.hasError = true;
+    }
+
+    if (transaction.cost < 0) {
+      errors.cost = "Cannot split a negative amount";
+      errors.hasError = true;
+    }
+
+    if (!transaction.users.byId[transaction.payed]) {
+      errors.payed = "Please choose a payee for this bill";
+      errors.hasError = true;
+    }
+
+    errors.users = {};
+    const uniqueNameSet = new Set();
+    for (const userId of transaction.users.allIds) {
+      const { name } = transaction.users.byId[userId];
+      if (!name) {
+        errors.users[userId] = "Please enter a name";
+        errors.hasError = true;
+      } else if (uniqueNameSet.has(name)) {
+        errors.users[userId] = "Same name can’t appear twice";
```

Comment:
Is there a reason you used 'RIGHT SINGLE QUOTATION MARK' (U+2019) here as opposed to 'APOSTROPHE' (U+0027)? Interested in how you ended up with it since the default is usually U+0027. I don't really have a preference either way unless someone else would like to comment.

Reply:
Yeah, I always preferred using the curly variants which looks more typographically correct, but now I wonder if it has the right semantics?

I double checked - it looks like Unicode version 13.0.0 recommends using U+2019 for contractions:

![image](https://user-images.githubusercontent.com/2609018/80158600-012b2600-861d-11ea-892c-69ce9c7b3472.png)

but [some people disagree with Unicode](https://tedclancy.wordpress.com/2015/06/03/which-unicode-character-should-represent-the-english-apostrophe-and-why-the-unicode-committee-is-very-wrong/) 🤔